src/DumpPersister.js

```diff
@@ -0,0 +1,116 @@
+const logger = require('./logging');
+const PromCollector = require('./metrics/PromCollector');
+const { saveEntryAssureUnique } = require('./store/dynamo');
+const initS3Store = require('./store/s3.js');
+const { asyncDeleteFile } = require('./utils/utils');
+const { RequestType } = require('./utils/utils');
+
+/**
+ *
+ */
+class DumpPersister {
+    /**
+     *
+     */
+    constructor({ tempPath, s3Config, disableFeatExtraction, workerPool }) {
+        this.tempPath = tempPath;
+        this.store = this.createDumpStorage(s3Config);
+        this.disableFeatExtraction = disableFeatExtraction;
+        this.workerPool = workerPool;
+    }
+
+    /**
+     * Initialize the service which will persist the dump files.
+     */
+    createDumpStorage(s3Config) {
+        if (s3Config?.region) {
+            return initS3Store(s3Config);
+        }
+        logger.warn('[DumpPersister] S3 is not configured!');
+    }
+
+    /**
+     * Persist the dump file to the configured store and save the  associated metadata. At the time of writing the
+     * only supported store for metadata is dynamo.
+     *
+     * @param {Object} sinkMeta - metadata associated with the dump file.
+     */
+    async persistDumpData(sinkMeta) {
+
+        // Metadata associated with a dump can get large so just select the necessary fields.
+        const { clientId } = sinkMeta;
+        let uniqueClientId = clientId;
+
+        // Because of the current reconnect mechanism some files might have the same clientId, in which case the
+        // underlying call will add an associated uniqueId to the clientId and return it.
+        uniqueClientId = await saveEntryAssureUnique(sinkMeta);
+
+        // Store the dump file associated with the clientId using uniqueClientId as the key value. In the majority of
+        // cases the input parameter will have the same values.
+        this.storeDump(clientId, uniqueClientId);
+    }
+
+    /**
+     * Store the dump to the configured store. The dump file might be stored under a different
+     * name, this is to account for the reconnect mechanism currently in place.
+     *
+     * @param {string} clientId - name that the dump file will actually have on disk.
+     * @param {string} uniqueClientId - name that the dump will have on the store.
+     */
+    async storeDump(clientId, uniqueClientId) {
+        const dumpPath = `${this.tempPath}/${clientId}`;
+
+        try {
+            await this.store?.put(uniqueClientId, dumpPath);
+        } catch (err) {
+            PromCollector.storageErrorCount.inc();
+
+            logger.error('Error storing: %s uniqueId: %s - %s', dumpPath, uniqueClientId, err);
+        } finally {
+            await asyncDeleteFile(dumpPath);
+        }
+    }
+
+    /**
+     *
+     * @param {*} meta
+     * @param {*} connectionInfo
+     */
+    processData(id, meta, connectionInfo) {
```

Comment:
This method feels like it should be part of a `DumpProcessor `class. Would it be hard to split it up?

Reply:
I started moving the things and found out that actually we are merging the two files. I'm open for refactrings. Maybe we can discuss tomorrow.